### ISSUE TITLE
ManyHasMany bug in ArrayMapper

### ISSUE THREAD

**Issue by martenb** [nextras/orm#507]
**Describe the bug**
When we upgrade from 3.1 to 4.0, our tests starts to fail in some special ManyHasMany relationships. I reproduce the error in nextras/orm tests.

**To Reproduce**
I will send PR with tests. I'm not sure, why it fails in mysql/etc, maybe another bug? :-)

**Versions::**
 - Database: ArrayMapper
 - Orm: 4.0
 - Dbal: 4.0

**Comment by martenb** [comment]
I have updated an publisher entity with tags property, the test looks like this

```php
	public function testManyHasMany(): void
	{
		// create 10 tags
		for ($id = 0; $id < 10; $id++) {
			$tag = $this->e(Tag::class, [
				'name' => (string) $id,
			]);
			$this->orm->tags->persistAndFlush($tag);
		}

		// assign all tags to publisher
		$publisher = $this->e(Publisher::class, [
			'tags' => $this->orm->tags->findAll()->fetchAll(),
		]);
		$this->orm->publishers->persistAndFlush($publisher);

		// assign publisher and only one tag
		$book = $this->e(Book::class, [
			'publisher' => $publisher,
			'tags' => [
				$this->orm->tags->getBy([]),
			],
		]);

		// so one tag in book
		Assert::same(1, $book->tags->count());

		$this->orm->books->persistAndFlush($book);

		// after pesist, there is all tags, not just one!
		Assert::same(1, $book->tags->count());
	}
```

then with ArrayMapper

![image](https://user-images.githubusercontent.com/13311472/112629134-2f6c4380-8e34-11eb-98ee-62701bab809d.png)

**Comment by hrach** [comment]
Thanks a lot, will take a look.

**Comment by hrach** [comment]
Fixed in master. I have cherry-picked the fixes to v4.0 - could you please check `v4.0.x-dev` composer version to verify it works for you? I will release stable version then. Thank you.

**Comment by martenb** [comment]
Looks ok, our tests now passing. Thanks a lot!

**Comment by hrach** [comment]
Thanks for checking out. Released as [v4.0.2](https://github.com/nextras/orm/releases/tag/v4.0.2).
